### PR TITLE
Use RawKernel without name_expressions to enable CuPy cache

### PR DESCRIPTION
**Pull request by mlazzarin**
In this PR I'm going to implement an alternative approach for kernel loading/compilation.
The main goal is to reduce compilation times.

Two considerations:
 - It seems like ``cp.RawKernel`` does not support templates.
 - By loading and compiling each kernel independently, we compile many times the helper methods / device functions that are called by many different kernels, so that if a circuit uses all different gates, the compilation time will be worse.

I tried the following approach:
 - Create a ``cp.RawModule`` for each kernel and each template specialization.
 - Create a lambda function that calls each rawmodule only when it's needed.

Dry run overhead: ~ 0.8 s (qft), ~ 0.65 s (supremacy, variational), ~ 0.5 s (only a one-qubit gate) against ~ 3.2 s with the main branch.
Notice that this results are very similar if we use CuPy 9.5.0 and CUDA 11.1.1, where the dry run overhead in the main branch is ~ 0.9 s.

It seems like the results are good, but I think that there are still margin of improvements. Also the readability of the code is worse, and we are using ``cp.RawModule`` for invidivual kernels as a workaround.
We may try to use ``cp.RawKernels`` and implement a template strategy by just manipulating strings.

Let me know your opinions.

### CODE REVIEW

**Comment by mlazzarin** [comment]
New implementation, now I use many ``cp.RawKernel`` and I implement the templates by manipulating the string with the code in Python. The CuPy cache is now working + we compile only what we need. Results:

<details><summary>qft</summary>

nqubits | Simulation time main | Simulation time smallkernels | Dry run overhead main | Dry run overhead smallkernels
-- | -- | -- | -- | --
3 | 0.00040 | 0.00036 | 3.1457420587539673 | 0.7802863359451294
4 | 0.00063 | 0.00058 | 3.1553324103355407 | 0.18701581954956054
5 | 0.00089 | 0.00083 | 3.1488601565361023 | 0.18863686323165893
6 | 0.00128 | 0.00116 | 3.155423128604889 | 0.19009333848953247
7 | 0.00164 | 0.00147 | 3.1571618914604187 | 0.19034855365753173
8 | 0.00199 | 0.00181 | 3.155499744415283 | 0.1883433699607849
9 | 0.00256 | 0.00231 | 3.157806408405304 | 0.19096494913101197
10 | 0.00319 | 0.00294 | 3.1664812922477723 | 0.19207983016967772
11 | 0.00391 | 0.00352 | 3.1634873151779175 | 0.19267864227294923
12 | 0.00466 | 0.00426 | 3.1650928258895874 | 0.1933570146560669
13 | 0.00534 | 0.00472 | 3.1504740357398986 | 0.19259796142578126
14 | 0.00624 | 0.00561 | 3.153474485874176 | 0.19510456323623657
15 | 0.00726 | 0.00649 | 3.14976806640625 | 0.1942536473274231
16 | 0.00781 | 0.00726 | 3.156639301776886 | 0.19579635858535765
17 | 0.00902 | 0.00800 | 3.157112193107605 | 0.1983433723449707
18 | 0.01022 | 0.00926 | 3.1660697937011717 | 0.19885637760162353
19 | 0.01171 | 0.01089 | 3.1653664469718934 | 0.20066726207733154
20 | 0.01529 | 0.01386 | 3.1783830523490906 | 0.20038068294525146
21 | 0.02195 | 0.02004 | 3.1642268657684327 | 0.20250802040100097
22 | 0.03282 | 0.03149 | 3.1735682010650637 | 0.2051405906677246
23 | 0.05514 | 0.05377 | 3.17226357460022 | 0.2042508602142334
24 | 0.10084 | 0.09936 | 3.1825984477996827 | 0.20787591934204103
25 | 0.19665 | 0.19602 | 3.1827134609222414 | 0.20782413482666015
26 | 0.40280 | 0.40062 | 3.177160406112671 | 0.20868186950683593
27 | 0.83474 | 0.83303 | 3.1749711990356446 | 0.21010322570800777
28 | 1.75560 | 1.75384 | 3.175797462463379 | 0.21109600067138667
29 | 3.69798 | 3.69570 | 3.1854668617248536 | 0.20963101387023908
30 | 7.82332 | 7.82032 | 3.1932161808013912 | 0.20800275802612322

</details>

<details><summary>variational</summary>

nqubits | Simulation time main | Simulation time smallkernels | Dry run overhead main | Dry run overhead smallkernels
-- | -- | -- | -- | --
3 | 0.00034 | 0.00031 | 3.138660216331482 | 0.346898627281189
4 | 0.00048 | 0.00042 | 3.1516257524490356 | 0.18903664350509644
5 | 0.00055 | 0.00046 | 3.148124563694 | 0.1897464394569397
6 | 0.00068 | 0.00058 | 3.1406522750854493 | 0.18845206499099731
7 | 0.00075 | 0.00066 | 3.1473780989646913 | 0.18753881454467775
8 | 0.00086 | 0.00073 | 3.154698443412781 | 0.18883769512176513
9 | 0.00096 | 0.00085 | 3.15038685798645 | 0.1889824390411377
10 | 0.00109 | 0.00098 | 3.157215893268585 | 0.18886677026748658
11 | 0.00122 | 0.00114 | 3.1425214409828186 | 0.18931114673614502
12 | 0.00139 | 0.00129 | 3.1518001317977906 | 0.19127660989761353
13 | 0.00147 | 0.00134 | 3.1463287711143493 | 0.18912984132766725
14 | 0.00164 | 0.00145 | 3.1549091219902037 | 0.19039055109024047
15 | 0.00170 | 0.00154 | 3.145152282714844 | 0.18988267183303834
16 | 0.00185 | 0.00164 | 3.143576610088348 | 0.19091131687164306
17 | 0.00197 | 0.00177 | 3.1475722193717957 | 0.19002219438552856
18 | 0.00236 | 0.00208 | 3.1435980558395387 | 0.19129763841629027
19 | 0.00305 | 0.00283 | 3.15607236623764 | 0.19024600982666015
20 | 0.00430 | 0.00418 | 3.1642513275146484 | 0.19223957061767577
21 | 0.00697 | 0.00663 | 3.151183032989502 | 0.19275822639465331
22 | 0.01238 | 0.01211 | 3.1775383949279785 | 0.19235563278198242
23 | 0.02308 | 0.02283 | 3.175465202331543 | 0.19030027389526366
24 | 0.04598 | 0.04565 | 3.172166633605957 | 0.1898524284362793
25 | 0.09227 | 0.09198 | 3.1634210109710694 | 0.19233784675598145
26 | 0.19068 | 0.19046 | 3.1652760028839113 | 0.19150023460388182
27 | 0.38924 | 0.38893 | 3.1548083305358885 | 0.1901273250579834
28 | 0.81020 | 0.80972 | 3.157725954055786 | 0.18795213699340818
29 | 1.66004 | 1.65945 | 3.163676309585571 | 0.18770236968994136
30 | 3.45552 | 3.45488 | 3.16283974647522 | 0.1775845527648925

</details>

<details><summary>supremacy</summary>

nqubits | Simulation time main | Simulation time smallkernels | Dry run overhead main | Dry run overhead smallkernels
-- | -- | -- | -- | --
3 | 0.00043 | 0.00036 | 3.150707507133484 | 0.18472249507904054
4 | 0.00053 | 0.00045 | 3.1542750120162966 | 0.1846892714500427
5 | 0.00067 | 0.00056 | 3.1555333018302916 | 0.1866701364517212
6 | 0.00080 | 0.00068 | 3.141358232498169 | 0.18656139373779296
7 | 0.00091 | 0.00078 | 3.1581531524658204 | 0.18657104969024657
8 | 0.00104 | 0.00090 | 3.140725791454315 | 0.18706315755844116
9 | 0.00112 | 0.00100 | 3.1619171380996702 | 0.186870276927948
10 | 0.00129 | 0.00118 | 3.142912137508392 | 0.18881402015686036
11 | 0.00149 | 0.00137 | 3.147163689136505 | 0.18911123275756836
12 | 0.00169 | 0.00150 | 3.148203730583191 | 0.18962695598602294
13 | 0.00177 | 0.00159 | 3.1470994472503664 | 0.18862301111221313
14 | 0.00193 | 0.00179 | 3.1393440127372743 | 0.18961775302886963
15 | 0.00207 | 0.00185 | 3.160055387020111 | 0.18933948278427123
16 | 0.00214 | 0.00199 | 3.146293020248413 | 0.18938645124435424
17 | 0.00233 | 0.00211 | 3.1468517065048216 | 0.1897303581237793
18 | 0.00285 | 0.00261 | 3.1564784049987793 | 0.19052075147628783
19 | 0.00382 | 0.00355 | 3.151102888584137 | 0.18810828924179077
20 | 0.00536 | 0.00507 | 3.144459939002991 | 0.190643310546875
21 | 0.00873 | 0.00840 | 3.154780149459839 | 0.19031567573547364
22 | 0.01560 | 0.01526 | 3.1629217147827147 | 0.19005827903747557
23 | 0.02921 | 0.02887 | 3.164820098876953 | 0.19114832878112792
24 | 0.05784 | 0.05751 | 3.1569369792938233 | 0.19093871116638184
25 | 0.11747 | 0.11710 | 3.1630221366882325 | 0.19100332260131836
26 | 0.24309 | 0.24279 | 3.1510163307189942 | 0.19039688110351563
27 | 0.49721 | 0.49678 | 3.169530248641968 | 0.19033470153808596
28 | 1.02472 | 1.02398 | 3.1642226696014406 | 0.18836970329284664
29 | 2.11980 | 2.11937 | 3.1576531410217283 | 0.18603715896606454
30 | 4.41254 | 4.41127 | 3.1763648033142093 | 0.17962164878845233

</details>

<details><summary>bv</summary>

nqubits | Simulation time main | Simulation time smallkernels | Dry run overhead main | Dry run overhead smallkernels
-- | -- | -- | -- | --
3 | 0.00032 | 0.00029 | 3.1418659925460815 | 0.48542282581329343
4 | 0.00042 | 0.00038 | 3.1464720487594606 | 0.18816239833831788
5 | 0.00052 | 0.00047 | 3.151164722442627 | 0.1886731505393982
6 | 0.00060 | 0.00055 | 3.1600461840629577 | 0.1872703790664673
7 | 0.00072 | 0.00063 | 3.155354642868042 | 0.18885055780410767
8 | 0.00083 | 0.00073 | 3.1485894918441772 | 0.1872439503669739
9 | 0.00095 | 0.00082 | 3.155393731594086 | 0.1869038701057434
10 | 0.00105 | 0.00095 | 3.146085774898529 | 0.18815128803253173
11 | 0.00120 | 0.00111 | 3.154426097869873 | 0.18852947950363158
12 | 0.00135 | 0.00121 | 3.157017970085144 | 0.188343608379364
13 | 0.00148 | 0.00131 | 3.147814393043518 | 0.18710625171661377
14 | 0.00157 | 0.00137 | 3.145067286491394 | 0.18876540660858154
15 | 0.00172 | 0.00152 | 3.155423128604889 | 0.18843706846237182
16 | 0.00187 | 0.00164 | 3.1450846672058104 | 0.19014612436294556
17 | 0.00186 | 0.00173 | 3.1600334525108336 | 0.1889096975326538
18 | 0.00232 | 0.00207 | 3.154076135158539 | 0.19392383098602295
19 | 0.00307 | 0.00291 | 3.1656222343444824 | 0.1901618003845215
20 | 0.00457 | 0.00431 | 3.158744955062866 | 0.1893435001373291
21 | 0.00745 | 0.00718 | 3.1617695808410646 | 0.19005942344665527
22 | 0.01323 | 0.01294 | 3.1526209831237795 | 0.18981122970581055
23 | 0.02515 | 0.02479 | 3.1683300018310545 | 0.18997573852539062
24 | 0.04978 | 0.04957 | 3.1479557514190675 | 0.19089384078979493
25 | 0.10131 | 0.10092 | 3.162819004058838 | 0.18861942291259765
26 | 0.20785 | 0.20761 | 3.14762282371521 | 0.19302873611450194
27 | 0.42850 | 0.42824 | 3.1510128021240233 | 0.18610920906066897
28 | 0.88521 | 0.88493 | 3.167058563232422 | 0.18783330917358398
29 | 1.82959 | 1.82921 | 3.1563172340393066 | 0.18466997146606445
30 | 3.78192 | 3.78125 | 3.1631487369537354 | 0.1768350601196289

</details>

<details><summary>qv</summary>

nqubits | Simulation time main | Simulation time smallkernels | Dry run overhead main | Dry run overhead smallkernels
-- | -- | -- | -- | --
3 | 0.00045 | 0.00039 | 3.1426882147789 | 0.18576139211654663
4 | 0.00084 | 0.00074 | 3.148140716552734 | 0.1906345844268799
5 | 0.00081 | 0.00072 | 3.1475600719451906 | 0.1884988784790039
6 | 0.00121 | 0.00108 | 3.1555628061294554 | 0.18863093852996826
7 | 0.00121 | 0.00110 | 3.1547466993331907 | 0.18788523674011232
8 | 0.00151 | 0.00136 | 3.1518974184989927 | 0.1896307110786438
9 | 0.00166 | 0.00141 | 3.1462293386459352 | 0.1906911015510559
10 | 0.00202 | 0.00186 | 3.1516823410987853 | 0.19165372848510742
11 | 0.00220 | 0.00192 | 3.1605575919151305 | 0.1930043339729309
12 | 0.00259 | 0.00237 | 3.1512955904006956 | 0.19252424240112304
13 | 0.00268 | 0.00235 | 3.1616410970687867 | 0.19283721446990967
14 | 0.00311 | 0.00265 | 3.1521570682525635 | 0.1934330701828003
15 | 0.00311 | 0.00274 | 3.160768473148346 | 0.19362775087356568
16 | 0.00349 | 0.00309 | 3.1496594667434694 | 0.19583089351654054
17 | 0.00347 | 0.00316 | 3.1649154305458067 | 0.19539986848831176
18 | 0.00452 | 0.00401 | 3.160015892982483 | 0.19349387884140015
19 | 0.00566 | 0.00522 | 3.1589847922325136 | 0.19695417881011962
20 | 0.00889 | 0.00849 | 3.1688802003860475 | 0.19663101434707642
21 | 0.01360 | 0.01313 | 3.1675274848937987 | 0.19628314971923827
22 | 0.02550 | 0.02508 | 3.172671604156494 | 0.19682583808898926
23 | 0.04657 | 0.04602 | 3.158849811553955 | 0.19658126831054687
24 | 0.09500 | 0.09454 | 3.1570056438446046 | 0.19880275726318358
25 | 0.18788 | 0.18722 | 3.1588647842407225 | 0.19809670448303224
26 | 0.39653 | 0.39630 | 3.1728618144989014 | 0.19995241165161132
27 | 0.79697 | 0.79612 | 3.1644497871398927 | 0.19733405113220215
28 | 1.70813 | 1.70710 | 3.169320821762085 | 0.1941514015197754
29 | 3.37040 | 3.36925 | 3.168869638442993 | 0.19393877983093244
30 | 7.28932 | 7.28826 | 3.1792341709136966 | 0.18935923576354963

</details>

**Comment by scarrazza** [comment]
Cool, thanks for the implementation results are quite good.

File: src/qibojit/custom_operators/backends.py [hunk 1]
```diff
@@ -192,28 +191,34 @@ def __init__(self):
...
-            kernels.append(f"apply_multi_qubit_gate_kernel{self.kernel_double_suffix[0:-2]}, {2**ntargets}>")
-            kernels.append(f"apply_multi_qubit_gate_kernel{self.kernel_float_suffix[0:-2]}, {2**ntargets}>")
-        kernels.append(f"collapse_state_kernel{self.kernel_double_suffix}")
-        kernels.append(f"collapse_state_kernel{self.kernel_float_suffix}")
-        kernels.append(f"initial_state_kernel{self.kernel_double_suffix}")
-        kernels.append(f"initial_state_kernel{self.kernel_float_suffix}")
+            kernels.append(f"{kernel}_kernel")
+            kernels.append(f"multicontrol_{kernel}_kernel")
+        kernels.append("collapse_state_kernel")
+        kernels.append("initial_state_kernel")
```
Comment (stavros11) [inline src/qibojit/custom_operators/backends.py:200]: Here we could avoid this loop (creating the `kernels` list) by loading the kernels and populating the `self.gates` dictionary directly. For example:
```Python
        # load core kernels
        self.gates = {}
        from qibojit.custom_operators import raw_kernels
        def kernel_loader(name, ktype):
            code = getattr(raw_kernels, name)
            code = code.replace("T", f"thrust::complex<{ktype}>")
            gate = cp.RawKernel(code, name, ("--std=c++11",))
            self.gates[f"{name}_{ktype}"] = gate

        for ktype in ("float", "double"):
            for name in self.KERNELS:
                kernel_loader(f"{name}_kernel", ktype)
                kernel_loader(f"multicontrol_{name}_kernel", ktype)
            kernel_loader("collapse_state_kernel", ktype)
            kernel_loader("initial_state_kernel", ktype)

        # load multiqubit kernels
        for ntargets in range(3, self.MAX_NUM_TARGETS + 1):
            for ktype in ("float", "double"):
                code = getattr(raw_kernels, "apply_multi_qubit_gate_kernel")
                code = code.replace("T", f"thrust::complex<{ktype}>")
                code = code.replace("nsubstates", str(2 ** ntargets))
                code = code.replace("MAX_BLOCK_SIZE", str(self.DEFAULT_BLOCK_SIZE))
                gate = cp.RawKernel(code, "apply_multi_qubit_gate_kernel", ("--std=c++11",))
                self.gates[f"apply_multi_qubit_gate_kernel_{ktype}_{ntargets}"] = gate
```
Feel free to keep whichever approach you prefer, given that none is significantly cleaner than the other.

File: src/qibojit/custom_operators/raw_kernels.py [hunk 2]
```diff
@@ -77,158 +83,249 @@ __device__ void _apply_two_qubit_gate(T& state0, T& state1, T& state2, T& state3
...
-  _apply_x<T>(state[i - tk], state[i]);
+  _apply_x(state[i - tk], state[i]);
 }
+"""
 
 
+multicontrol_apply_y_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_y}
+{multicontrol_index}"""+"""
```
Comment (stavros11) [inline src/qibojit/custom_operators/raw_kernels.py:278]: Do we really need the string addition (`+`) here? Wouldn't something like:
```Python
multicontrol_apply_y_kernel = f"""
#include <cupy/complex.cuh>
{_apply_y}
{multicontrol_index}
// C++ implementation of gates.py:multicontrol_apply_y_kernel()
extern "C"
___global___ void ...
"""
```
have the same effect?

**Review by stavros11 — APPROVED** [review]
Thanks for implementing this and the benchmarks. Looks good and dry run performance will be much better for average usage, given that a small subset of kernels are typically used by a single script so there is no need to compile or load all. Good news that the cache is also working without the template.

If we intentionally write a script that calls all kernels (worst case), the dry run goes back to 2sec (with cache), right?

File: src/qibojit/custom_operators/raw_kernels.py (hunk 2 above)
Comment (mlazzarin) [reply src/qibojit/custom_operators/raw_kernels.py:278]: It's a concatenation between a f-string and a standard one. With your suggestion, the body of the kernel within brackets would be considered as Python code to be evaluated.

File: src/qibojit/custom_operators/backends.py (hunk 1 above)
Comment (mlazzarin) [reply src/qibojit/custom_operators/backends.py:200]: Approved, thanks.

File: src/qibojit/custom_operators/raw_kernels.py (hunk 2 above)
Comment (stavros11) [reply src/qibojit/custom_operators/raw_kernels.py:278]: I see, you are right. Thank you for the explanation and updates.

**Comment by mlazzarin** [comment]
> If we intentionally write a script that calls all kernels (worst case), the dry run goes back to 2sec (with cache), right?

I don't know about calling all kernels, but if I manually compile all kernels:
```
        # load core kernels
        self.gates = {}
        from qibojit.custom_operators import raw_kernels
        def kernel_loader(name, ktype):
            code = getattr(raw_kernels, name)
            code = code.replace("T", f"thrust::complex<{ktype}>")
            gate = cp.RawKernel(code, name, ("--std=c++11",))
            self.gates[f"{name}_{ktype}"] = gate
            self.gates[f"{name}_{ktype}"].compile()

        for ktype in ("float", "double"):
            for name in self.KERNELS:
                kernel_loader(f"{name}_kernel", ktype)
                kernel_loader(f"multicontrol_{name}_kernel", ktype)
            kernel_loader("collapse_state_kernel", ktype)
            kernel_loader("initial_state_kernel", ktype)

        # load multiqubit kernels
        name = "apply_multi_qubit_gate_kernel"
        for ntargets in range(3, self.MAX_NUM_TARGETS + 1):
            for ktype in ("float", "double"):
                code = getattr(raw_kernels, name)
                code = code.replace("T", f"thrust::complex<{ktype}>")
                code = code.replace("nsubstates", str(2 ** ntargets))
                code = code.replace("MAX_BLOCK_SIZE", str(self.DEFAULT_BLOCK_SIZE))
                gate = cp.RawKernel(code, name, ("--std=c++11",))
                self.gates[f"{name}_{ktype}_{ntargets}"] = gate
                self.gates[f"{name}_{ktype}_{ntargets}"].compile()
```
the import time increases from ~ 1.15 s to ~ 1.3 s, while the dry run overhead decreases to ~ 0.1 s.
If I manually remove the cache, the import time is ~ 10 s.
So, we have two different points:
 - The CuPy cache works, while previously it wasn't working (I remember a decrease of ~ 0.5 s due to caching, maybe it worked for only some portions of the kernels).
 - Once the cache is done, loading all kernels during import or only the called ones during execution is kind of equivalent.
So, it looks like the real advantage here is the cache, not the single kernel loading.

By the way, there is a WIP PR in cupy regarding the implementation of caching for RawModule with templates.
Such feature would fix the dry run problem also with our current implementation in the main branch.

**Comment by mlazzarin** [comment]
Before merging this PR, there is an open pull request in the cupy repo which may implement the caching system also with RawModule with ``name_expressions``. If everything goes well, we don't need to change anything from the current main branch of qibojit. https://github.com/cupy/cupy/pull/6159

**Comment by mlazzarin** [comment]
Concerning this PR, it should be complete. However, it's been a while since it's been written and approved, and a lot as changed since then. The merge commit was flawless, and everything looks good to me, but I'm still worried about potential regressions or conflicts, so please review this PR again.

Tests are fine on my laptop (nvidia gpu), I didn't check on AMD because I don't have access to an idle gpu right now.

**Comment by scarrazza** [comment]
@mlazzarin I am getting some compilation errors:
```text
E           cupy.cuda.compiler.CompileException: <built-in>:1:10: fatal error: '__clang_hip_runtime_wrapper.h' file not found
E           #include "__clang_hip_runtime_wrapper.h"
E                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
E           1 error generated when compiling for gfx906.
E           Error: Failed to compile opencl source (from CL or HIP source to LLVM IR).

../.local/lib/python3.8/site-packages/cupy/cuda/compiler.py:671: CompileException
```

The card is now available if prefer to give a try.

**Review by stavros11 — APPROVED** [review]
Thanks for updating this, looks good to me and tests (qibo and qibojit) are passing on my machine. I do not have access to AMD either so cannot test that one.

There is significant improvement in import time for me, for example from 4.5sec to 1.2sec on qibomachine. Have you done any extensive benchmarks comparing this with main for all times we log (import, dry run, simulation)? In qibomachine for 25-qubit QFT (using the latest compare.py script) I get:

--  | main | smallkernels  
-- | -- | --
import time | 4.411 | 1.189
dry run time | 0.403 | 0.617
simulation time | 0.289 | 0.288

Generally I get a big decrease in import time and a small increase in dry run by switching to this branch. This is from some quick runs, so it may not be very accurate but seems to be consistent on both qibomachine and dgx. Either way, I think what we have in this branch is preferable.

**Comment by mlazzarin** [comment]
> Generally I get a big decrease in import time and a small increase in dry run by switching to this branch. This is from some quick runs, so it may not be very accurate but seems to be consistent on both qibomachine and dgx. Either way, I think what we have in this branch is preferable.

This is expected because in the ``main`` branch we compile import time, while here we compile the first time each kernel is called. We can go for both option regardless of using template or not.

**Comment by mlazzarin** [comment]
> 

Tests are fine also on AMD. What are you trying to run exactly?

**Comment by mlazzarin** [comment]
> @mlazzarin I am running pytest-3 inside the qibojit folder, are you doing the same?

Yes, I'm running ``pytest`` installed with pip. I guess it should be the same.
I'm using cupy v9.5.0 with rocm 4.3.

Do the tests work fine for you with the ``main`` branch?

**Review by andrea-pasquale — APPROVED** [review]
Thanks for this implementation. Looks good to me.

**Review by scarrazza — APPROVED** [review]

**Comment by stavros11** [comment]
Here are some multigpu benchmarks using this branch vs master:

<details>
<summary>qft</summary>

| nqubits (accelerators)   |   dry run master | simulation master |   dry run smallkernels |   simulation smallkernels |
|:-------------------------|----------------------:|-------------------------------:|----------------------------:|-------------------------------------:|
| 31 (2/GPU:3)             |               78.275  |                        67.0925 |                     73.5205 |                              66.8976 |
| 31 (1/GPU:2+1/GPU:3)     |               66.0005 |                        46.2453 |                     60.5393 |                              45.3707 |
| 32 (4/GPU:3)             |              202.643  |                       189.696  |                    198.299  |                             189.924  |
| 32 (2/GPU:2+2/GPU:3)     |              148.502  |                       130.196  |                    148.687  |                             130.293  |

</details>
<details>
<summary>variational</summary>

| nqubits (accelerators)   |   dry run master | simulation master |   dry run smallkernels |   simulation smallkernels |
|:-------------------------------------|----------------------:|-------------------------------:|----------------------------:|-------------------------------------:|
| 31 (2/GPU:3)                         |               65.7648 |                        55.2782 |                     63.6881 |                              55.209  |
| 31 (1/GPU:2+1/GPU:3)                 |               58.0932 |                        37.6242 |                     52.2991 |                              38.4544 |
| 32 (4/GPU:3)                         |              177.011  |                       164.179  |                    173.105  |                             164.666  |
| 32 (2/GPU:2+2/GPU:3)                 |              132.467  |                       115.142  |                    130.772  |                             114.212  |
| 32 (1/GPU:0+1/GPU:1+1/GPU:2+1/GPU:3) |              120.622  |                        81.5669 |                    114.192  |                              80.9547 |

</details>
<details>
<summary>supremacy</summary>

| nqubits (accelerators)   |   dry run master | simulation master |   dry run smallkernels |   simulation smallkernels |
|:-------------------------------------|----------------------:|-------------------------------:|----------------------------:|-------------------------------------:|
| 31 (2/GPU:3)                         |               65.8643 |                        56.7581 |                     64.296  |                              58.3684 |
| 31 (1/GPU:2+1/GPU:3)                 |               59.2494 |                        39.3849 |                     49.0412 |                              39.1617 |
| 32 (4/GPU:3)                         |              179.505  |                       169.539  |                    175.384  |                             167.733  |
| 32 (2/GPU:2+2/GPU:3)                 |              133.158  |                       116.971  |                    130.873  |                             117.167  |
| 32 (1/GPU:0+1/GPU:1+1/GPU:2+1/GPU:3) |              127.176  |                        94.4546 |                    118.036  |                              95.0658 |

</details>

Indeed, there is some improve in the dry run, though not as big as I expected. Dry run is now more competitive with qibotf though.

**Pull request merged**